### PR TITLE
Music Rental

### PR DESCRIPTION
**Pull request by BSeiff**
I did the Music Rental biz-scenario for homework. Kindly review.

### CODE REVIEW

File: 1 table.sql [hunk 1]
```diff
@@ -0,0 +1,30 @@
+use MusicNoteDB 
+go 
+
+drop table if exists MusicRental
+go 
+create table dbo.MusicRental(
+    InstrumentId int not null identity primary key,
```
Comment (AvigailGellis) [inline 1 table.sql:7]: The pk should be the name of the table. Either name your table Instrument or your pk MusicRentalId

File: 1 table.sql [hunk 2]
```diff
@@ -0,0 +1,30 @@
+use MusicNoteDB 
+go 
+
+drop table if exists MusicRental
+go 
+create table dbo.MusicRental(
+    InstrumentId int not null identity primary key,  
+    CustomerFirstName varchar(30) not null constraint c_MusicNote_CustomerFirstName_cannot_be_blank check (CustomerFirstName > ''),
+    CustomerLastName varchar(30) not null constraint c_MusicNote_CustomerLastName_cannot_be_blank check (CustomerLastName > ''),
+    Instrument varchar(20) not null constraint c_MusicNote_Instrument_cannot_be_blank check (Instrument > ''),
```
Comment (AvigailGellis) [inline 1 table.sql:10]: Nice that you made this constraint, but the owner did not say he is limited to these 4 categories, happens to be those are the only ones in his data so far. Rather add constraint that can't be blank.

File: 1 table.sql [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+    Category varchar(14) not null constraint c_MusicNote_Category_is_Brass_Wind_Percussion_or_String check(Category in ('Brass', 'Wind', 'Percussion', 'String')),
```
Comment (AvigailGellis) [inline 1 table.sql:11]: Nice that you made this constraint, but the owner did not say he is limited to these 4 categories, happens to be those are the only ones in his data so far. Rather add constraint that can't be blank.

File: 1 table.sql [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+    MonthlyRentalFee decimal(4,2) not null constraint c_MusicNote_MonthlyRentalFee_greater_than_0 check (MonthlyRentalFee > 0),
+    DateRented date not null constraint c_MusicNote_DateRented_cannot_be_blank check(DateRented > '') ,
```
Comment (AvigailGellis) [inline 1 table.sql:13]: No need for blank constraint on date, DateRented should  be before the current date.

File: 1 table.sql [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    DateReturned date null,
```
Comment (AvigailGellis) [inline 1 table.sql:14]: DateReturned cannot be in the future.

File: 1 table.sql [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    UniqueInstrumentId as concat(substring(Instrument,1,3), lower(substring(Category, 1,3))),
+    TotalProfit as case 
+        when DateReturned is not null then datediff(month, DateRented, DateReturned) * MonthlyRentalFee
+        else 0
+        end persisted,
+    DateInserted datetime default getdate(),
```
Comment (AvigailGellis) [inline 1 table.sql:20]: Don't allow nulls for DateInserted

File: 1 table.sql [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    constraint c_MusicNote_DateReturned_after_DateRented check(DateRented < DateReturned),
+    constraint c_MusicNote_DateRented_to_DateReturned_min_1_month_and_max_12_months check((DateReturned is not null and datediff(month, DateRented, DateReturned) between 1 and 12) or (DateReturned is null)),
+    constraint c_MusicNote_DateRented_cannot_be_future_date check(DateRented <= getdate())
```
Comment (AvigailGellis) [inline 1 table.sql:23]: This should be a constraint directly on the column itself.

File: 2 data.sql [hunk 8]
```diff
@@ -0,0 +1,19 @@
+use MusicNoteDB
+go 
+
+delete MusicRental
+
+insert MusicRental(CustomerFirstName, CustomerLastName, Instrument, Category, MonthlyRentalFee, DateRented, DateReturned)
+select 'Esther', 'Shields', 'Flute', 'Wind', 30, 'Jan 01 2021', 'May 03, 2021'
```
Comment (AvigailGellis) [inline 2 data.sql:7]: Improvement: It would be better if dates were in 'yyyy-mm-dd' format

File: 1 table.sql (hunk 1 above)
Comment (BSeiff) [reply 1 table.sql]: Primary key renamed MusicRental

File: 1 table.sql (hunk 3 above)
Comment (BSeiff) [reply 1 table.sql]: Changed constraint to blank

File: 1 table.sql (hunk 4 above)
Comment (BSeiff) [reply 1 table.sql]: Removed this and replaced it with DateRented cannot be future date

File: 1 table.sql (hunk 5 above)
Comment (BSeiff) [reply 1 table.sql]: Added this constraint

File: 1 table.sql (hunk 6 above)
Comment (BSeiff) [reply 1 table.sql]: Added not null

File: 1 table.sql (hunk 7 above)
Comment (BSeiff) [reply 1 table.sql]: Moved. See above comment.

File: 2 data.sql (hunk 8 above)
Comment (BSeiff) [reply 2 data.sql]: Date format changed to the suggested format

**Review by BSeiff — COMMENTED** [review]
Thanks for reviewing. Changes were hopefully implemented properly, please see new commit.

**Comment by AvigailGellis** [comment]
Amazing job! 100%

**Pull request closed without merging**